src/tox/session/commands/run/parallel.py

```diff
@@ -76,25 +83,56 @@ def run_in_thread(tox_env, os_env):
                     reporter.quiet(message)
 
         threads = []
+        processes = {}
         todo_keys = set(venv_dict.keys())
         todo = OrderedDict((n, todo_keys & set(v.envconfig.depends)) for n, v in venv_dict.items())
         done = set()
-        while todo:
-            for name, depends in list(todo.items()):
-                if depends - done:
-                    # skip if has unfinished dependencies
-                    continue
-                del todo[name]
-                venv = venv_dict[name]
-                semaphore.acquire(blocking=True)
-                spinner.add(name)
-                thread = Thread(target=run_in_thread, args=(venv, os.environ.copy()))
-                thread.start()
-                threads.append(thread)
-            if todo:
-                # wait until someone finishes and retry queuing jobs
-                finished.wait()
-                finished.clear()
+        try:
+            while todo:
+                for name, depends in list(todo.items()):
+                    if depends - done:
+                        # skip if has unfinished dependencies
+                        continue
+                    del todo[name]
+                    venv = venv_dict[name]
+                    semaphore.acquire(blocking=True)
+                    spinner.add(name)
+                    thread = Thread(
+                        target=run_in_thread, args=(venv, os.environ.copy(), processes)
+                    )
+                    thread.daemon = True
+                    thread.start()
+                    threads.append(thread)
+                if todo:
+                    # wait until someone finishes and retry queuing jobs
+                    finished.wait()
+                    finished.clear()
+            for thread in threads:
+                while thread.is_alive():
+                    thread.join(0.05)
+                    # join suspends signal handling (ctrl+c), periodically time-out to check for it
+        except KeyboardInterrupt:
+            reporter.verbosity0("keyboard interrupt parallel - stopping children")
+            while True:
+                # do not allow to interrupt until children interrupt
+                try:
+                    # first try to request a graceful shutdown
+                    for name, proc in list(processes.items()):
+                        if proc.returncode is None:
+                            proc.send_signal(
+                                signal.CTRL_C_EVENT if sys.platform == "win32" else signal.SIGINT
+                            )
+                            reporter.verbosity2("send CTRL+C {}-{}".format(name, proc.pid))
+                    if len(threads):
+                        threads[0].join(1.2)  # wait at most 200ms for all to finish
```

Comment:
The code and the comments disagree: 1.2 seconds != 200 ms.

Reply:
Yeah, still struggling to make this pass on Windows, will fix once I succeed. 👍